src/rgw/rgw_bucket_sync_cache.h

```diff
@@ -23,31 +23,34 @@ namespace rgw::bucket_sync {
 // per bucket-shard state cached by DataSyncShardCR
 struct State {
   // the source bucket shard to sync
-  rgw_bucket_shard key;
+  std::pair<rgw_bucket_shard, uint64_t> key;
```

Comment:
TODO: introduce a `rgw_bucket_log_shard` struct that includes the generation

i avoided this before, because the optional gen was specific to data sync paths, so couldn't be generalized this way

but changing this would cause lots of backport conflicts that i don't want to deal with, so we'll hold off on it for now